### PR TITLE
br: make lightning tidb backend compatible with mysql

### PR DESCRIPTION
**Pull request by glorv**
### What problem does this PR solve?
Allow use tidb-lightning tidb-backend to load data into other mysql compatible database. This is the last step to replace loader with tidb-lightning.

### What is changed and how it works?
- Add a `ParseServerInfo` method the extract server type and version info from `SELECT version()` result
- Make some tidb specified steps optional and only do them with local/importer backend

### Check List 

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
 - Manual run some of the integration test for tidb backend with MySQL downstream
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
Make tidb-lightning tidb-backend compatible with MySQL
```

### CODE REVIEW

File: br/pkg/lightning/common/util.go [hunk 1]
```diff
@@ -20,6 +20,7 @@ import (
 	"encoding/json"
 	stderrors "errors"
 	"fmt"
+	"github.com/pingcap/tidb/br/pkg/utils"
```
Comment (kennytm) [inline br/pkg/lightning/common/util.go:23]: move this to next group

File: br/pkg/lightning/restore/tidb.go [hunk 2]
```diff
@@ -53,6 +52,13 @@ var defaultImportantVariables = map[string]string{
 	"group_concat_max_len":    "1024",
 }
 
+// defaultImportVariablesTiDB is used in ObtainImportantVariables to retrieve the system
+// variables from downstream in local/importer backend. The values record the default
+//// values if missing.
```
Comment (kennytm) [inline br/pkg/lightning/restore/tidb.go:57]:
```suggestion
// values if missing.
```

File: br/pkg/lightning/restore/tidb.go [hunk 3]
```diff
@@ -305,18 +319,24 @@ func UpdateGCLifeTime(ctx context.Context, db *sql.DB, gcLifeTime string) error
...
+func ObtainImportantVariables(ctx context.Context, g glue.SQLExecutor, needTiDBVars bool) map[string]string {
 	var query strings.Builder
 	query.WriteString("SHOW VARIABLES WHERE Variable_name IN ('")
 	first := true
 	for k := range defaultImportantVariables {
 		if first {
 			first = false
 		} else {
-			query.WriteString("','")
+			query.WriteByte(',')
```
Comment (kennytm) [inline br/pkg/lightning/restore/tidb.go:330]: ???????

Comment (glorv) [reply br/pkg/lightning/restore/tidb.go]: my mistake😂

File: br/pkg/version/version.go [hunk 4]
```diff
@@ -238,3 +239,72 @@ func NormalizeBackupVersion(version string) *semver.Version {
...
+	serverInfo := ServerInfo{}
+	switch {
+	case strings.Contains(lowerCase, "tidb"):
+		serverInfo.ServerType = ServerTypeTiDB
+	case strings.Contains(lowerCase, "mariadb"):
+		serverInfo.ServerType = ServerTypeMariaDB
+	case mysqlVersionRegex.MatchString(lowerCase):
+		serverInfo.ServerType = ServerTypeMySQL
+	default:
+		serverInfo.ServerType = ServerTypeUnknown
```
Comment (kennytm) [inline br/pkg/version/version.go:291]: i think we can just treat `ServerTypeUnknown` as the same as `ServerTypeMySQL`.

File: br/pkg/lightning/restore/restore.go [hunk 5]
```diff
@@ -1552,6 +1549,9 @@ func (rc *Controller) switchToImportMode(ctx context.Context) {
 }
 
 func (rc *Controller) switchToNormalMode(ctx context.Context) error {
+	if rc.isTiDBBackend() {
+		return nil
+	}
```
Comment (kennytm) [inline br/pkg/lightning/restore/restore.go:1554]: place this inside `switchTiKVMode`

**Review by kennytm — APPROVED** [review]
rest LGTM

File: br/pkg/lightning/restore/restore.go [hunk 6]
```diff
@@ -1451,9 +1440,17 @@ func (tr *TableRestore) restoreTable(
...
+		versionStr, err := version.FetchVersion(ctx, db)
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+
+		versionInfo := version.ParseServerInfo(versionStr)
 
 		// "show table next_row_id" is only available after v4.0.0
-		if tidbVersion.Major >= 4 && (rc.cfg.TikvImporter.Backend == config.BackendLocal || rc.cfg.TikvImporter.Backend == config.BackendImporter) {
+		if versionInfo.ServerType == version.ServerTypeTiDB && versionInfo.ServerVersion.Major >= 4 &&
```
Comment (Little-Wallace) [inline br/pkg/lightning/restore/restore.go:1452]: It seems that if ServerType does not equal `ServerTypeTiDB`, this backend can not be `BackendLocal` and `BackendImporter`.  We can check it in config

File: br/pkg/lightning/restore/restore.go [hunk 7]
```diff
@@ -1667,6 +1667,10 @@ func (rc *Controller) enforceDiskQuota(ctx context.Context) {
 }
 
 func (rc *Controller) setGlobalVariables(ctx context.Context) error {
+	// skip for tidb backend to be compatible with MySQL
+	if rc.isTiDBBackend() {
```
Comment (Little-Wallace) [inline br/pkg/lightning/restore/restore.go:1671]: Could we rename `BackendTiDB` to `BackendSQL` ?

Comment (glorv) [reply br/pkg/lightning/restore/restore.go:1671]: `BackendSQL` should be a better name. Not sure if we should also change the corresponding config fields. @kennytm what do you think?

File: br/pkg/version/version.go (hunk 4 above)
Comment (glorv) [reply br/pkg/version/version.go:291]: This logic is copy from dumpling, after we merge dumpling into tidb repo in this sprint, we can move them into a common package to reduce duplicate code.😅

File: br/pkg/lightning/restore/restore.go (hunk 7 above)
Comment (kennytm) [reply br/pkg/lightning/restore/restore.go:1672]: yes, but we could do it in another PR.

`backend = "tidb"` in the config should still be accepted (as alias of `backend = "sql"`).

**Comment by glorv** [comment]
@Little-Wallace @gozssky PTAL again, thanks~

**Comment by gozssky** [comment]
Can you add an integration test for MySQL? (seems not easy 😂) Or do some manual test and paste the steps in PR description.

**Comment by glorv** [comment]
> Can you add an integration test for MySQL? (seems not easy 😂) Or do some manual test and paste the steps in PR description.

It's not easy to add integration test with MySQL. We can depend on DM's integration test after https://github.com/pingcap/dm/pull/2051 this PR has merged.😅

**Review by gozssky — APPROVED** [review]

**Pull request merged**